### DISCUSSION TITLE
How to run the Directus app in development mode?

### DISCUSSION THREAD

**Discussion by mpenna** [directus/directus#17967] [Archive]
I've been trying to achieve this for days already, but to no avail: I can't figure out how to get the Directus app to run in 'development' mode. I've read somewhere that this is only possible when running the project locally, but I have already tried that route but couldn't get it to run in dev mode. The idea is to enable Vue's dev tools, which would be of great help when developing extensions. I'd much appreciate any pointers on how to get this ball rolling.

**Comment by Zack-Heisnberg** [comment]
What the problems you are facing when you run the source code version ? follow this guide 
https://docs.directus.io/contributing/running-locally.html

**Comment by mpenna** [comment]
Thanks for your reply. I just can't put it into "dev" mode. Browser's console keeps indicating that it is running in "production" mode, thus disabling Vue's dev tools altogether.

I have indeed followed the referred documentation, but I always end up with the same scenario. Setting the following env vars inside /api/.env seems to make no difference:

```md
SERVE_APP=true
NODE_ENV=development
```

**Reply by br41nslug** [reply]
As Zack mentioned running the App locally in dev mode is documented here: https://docs.directus.io/contributing/running-locally.html#_7-start-the-development-server
Did you try running the app in dev mode `pnpm --filter @directus/app dev`?

**Reply by mpenna** [reply]
Following the "multiple packages" strategy stated in the docs, this is what I got so far. 

`pnpm --filter directus dev` running on one terminal
`pnpm --filter @directus/app dev` running on a second terminal

Set `SERVE_APP=false` inside `/api/.env`.  

I run everything behind a reverse proxy (Traefik), and I use a locally resolved domain name for all services, everything configured with SSL certificates (Let's Encrypt). Everything is working fine, except Directus (running from source), but only in dev mode. Normal Directus instance works just fine, but always in production mode.

Attached are two screen captures, showing how API and APP domains are showing in the browser. 

![Screenshot from 2023-03-28 15-32-09](https://user-images.githubusercontent.com/6953141/228334770-d9b6fa36-fd8b-416b-a3c3-37b2ac9c4370.png)

![Screenshot from 2023-03-28 15-32-18](https://user-images.githubusercontent.com/6953141/228334825-ab6a513b-d0e5-41cd-a5a1-37d74a0e69a2.png)

**Reply by br41nslug** [reply]
Ah yeah development is intended for local use. As the name may indicate 😬 You should not run that through a proxy in dev mode or connect it to the internet at all for that matter.

**Reply by mpenna** [reply]
But I am indeed running everything locally and dockerized. The https domains are just a way to mimick an eventual production hosting scenario.

**Reply by br41nslug** [reply]
But (in development mode) running a HMR websocket and file access to your local system and all that insecure development goodness is not representative at all of a production environment. Running the provided docker should be a lot closer match to an eventual production setup.

**Reply by mpenna** [reply]
Sure, I tend to agree, but I do it this way because this architecture is one step closer to a prod scenario.

Setting the `server.host: "0.0.0.0"` property inside /app/vite.config.js` apparently did the trick. Browsing the admin panel is somewhat slower, but now Vue's dev tools is available.

![Screenshot from 2023-03-28 15-53-42](https://user-images.githubusercontent.com/6953141/228340065-7922c46f-6c8c-4763-8be3-be3a96cb65b4.png)

**Reply by br41nslug** [reply]
Awesome you got it working :)

**Comment by emahuni** [comment]
I think Directus should make this simple. I am using self-hosted version, where I install using pnpm. Directus team should allow the App to run in development mode once the /api/.env file is set to NODE_ENV=development. This allows way better development experience without forking the repo to run the APP as a separate process.

**Comment by emahuni** [comment]
The way I did it was to fork the APP and then clone the repo using git sparse, just to get to what I wanted. Then change all package.json workspace:* versions to their actual npm available packages, install dependencies, change API_PATH to point to my self-hosted api dir, and run the app using the dev script as if I am developing it. That way I have no issues with it. 

I am doing all this so that I can have devtools, but I just wish self-hosted would just work easily, where I install everything as a npm package and then run using `pnpm directus start` would just work if I add a `NODE_ENV=development` env var.

**Reply by emahuni** [reply]
The way I handle and structure my projects makes it easier to deal with this problem:

I create my projects with the following structure:
![image](https://github.com/user-attachments/assets/b28a57a0-f3c4-41e8-8674-722414203da8)

`/package.json` contents:
```json
{
  "name": "@emahuni/project.monorepo",
  "version": "1.0.0",
  "scripts": {
    "build:exts": "pnpm --recursive --no-bail --parallel --filter \"*/directus-extension-*\" --filter \"directus-extension-*\" run build",
    "dev:exts": "pnpm --recursive --no-bail --parallel --filter \"*/directus-extension-*\" --filter \"directus-extension-*\" run dev",
    "directus:start": "echo \"\\033[33m\\nStarting production API, ensure Nginx /admin location conf has correct port...\\n\\033[0m\" && pnpm -C ./api run start"
  }
}
```

`pnpm-workspace.yaml` contents:
```yaml
packages:
  # self-hosted node module based directus api and extensions
  - api/
  - api/extensions/**/*
  # project apps, may or may not use the api
  - apps/*
  # clone of directus packages that need to be run locally or modified 
  - apps/directus/**/*
  
  # things i didn't include in the screenshot and in package.json scripts etc below
  # project constants
  - constants/
  # project types
  - types/
  # allow separate management of other independent non-directus packages. Extensions that depend on these will find them missing if built at the same time, hence the separation 
  - packages/libraries/*
 # manage directus extensions separately
  - packages/directus.extensions/*
```


`/api/package.json` contents: 
```json
{
  "name": "@emahuni/project.api",
  "version": "1.0.0",
  "description": "Description for API",
  "main": "index.js",
  "scripts": {
    "start": "directus start",
    "dev": "SERVE_APP=false NODE_ENV=development directus start",
    "test": "vitest"
  },
  "dependencies": {
    "directus": "^11.5.1",
    "pg": "^8.13.1"
  },
}
```

also add a `.env` file in the api dir. I didn't put it in the screenshot, but it should define your environment according to Directus  documentation: https://docs.directus.io/self-hosted/config-options.html

Now let's clone the Directus repo, but sparsley.
```bash
cd apps
git clone --no-checkout --filter=blob:none https://github.com/directus/directus.git
cd directus
git sparse-checkout init --no-cone
git sparse-checkout set "app/*"
git checkout -b my-project-app-branch
```
This will only clone the directus app dir and branch to `my-project-app-branch` to make it easier for future pulls etc. You should have something like this as your dir structure:
![image](https://github.com/user-attachments/assets/601f5f5f-31b7-465b-9baf-d0484da6091d)

Change the devDependencies in `/apps/directus/app/package.json` workspaced packages as follows:
![image](https://github.com/user-attachments/assets/8e686ec1-cf1e-447e-9d9a-68a784cbe4b1)

To update this in the future just `git pull` then commit and merge as necessary keeping your essential changes to keep this hack running.

One last hack:
Edit this line in `/apps/directus/app/vite.config.js`as follows (this makes sure your local api extensions are found):
![image](https://github.com/user-attachments/assets/657950bb-7781-4f25-8224-adf3a3e4b630)

At this point you are ready to install:
```bash
pnpm i
```
This will install the required dependencies for the api and the directus app that you cloned.

Now cd to your api dir, and run `pnpm directus bootstrap` to create db tables etc and `pnpm dev` to run the api without the app studio.

Open another terminal window and cd to the `/apps/directus/app` dir and run `pnpm dev`, this will create a new vite dev server and you use that to work with your app, that will have vue dev tools etc. It will run directus in development mode. Your local extensions will be found without any issues...

**Reply by emahuni** [reply]
why i do this instead of docker, I like to peep in the app dir and see how things are happening under the hood and create my own extensions that follow Directus' philosophies... plus I can quickly modify how something works in the app studio and see how I can do certain things I wouldn't otherwise be able to do running with Docker, besides the straight forward path to devtools 😉.

**Reply by emahuni** [reply]
That said, I totally understand why Directus team doesn't offer the development env for the app, by the time you run it in self-hosted mode, you are getting the compiled production version of it, hence the clonning I am doing here. I think the only part that is preventing this working without much modification is the vite.config line, if only that was env configurable then it was just a matter of cloning it and running the app as @falmanna indicated.

**Reply by emahuni** [reply]
If you don't like my structure or the process is too long etc, just clone the whole directus repo somewhere, then do the last part where you edit that `vite.config` line to point to where your actual self-hosted api is located, install deps and run the app with vite `pnpm --filter @directus/app dev`. Just to keep things lean on your resources, set `SERVE_APP=false` in .env file, but this doesn't really matter, the dev app works alongside the selfhosted one very well.

**Comment by falmanna** [comment]
I found a really easy way, just turn off the app in your repo by setting `SERVE_APP=false` then clone directus repo and start the app in dev mood separately using `pnpm --filter @directus/app dev`. This way I am still running my full directus setup with extensions and data, while having the app run in dev mood.

Obviously make sure that you are running both on the same version to not have problems.  

So far seems to be working fine.

**Reply by falmanna** [reply]
Ops, seems like my extension is not getting registered.

**Reply by emahuni** [reply]
The reason why it's not registering extension is exactly why I said you've to edit a few things in the app dir itself. 
This tells the app where your api is running from and registers the extensions correctly.

The vite.cofig.js file has changed from the time I wrote that comment. There is less things to change in it, which is a good direction towards doing this easily.

Let me detail it how to do it on my above comment so others understand how to do it.

**Comment by xenobytezero** [comment]
So just to confirm, if I am just developing extensions, and I'm running Directus in Docker, there is no way to switch Directus to development mode, meaning I can't use hot reload, or use the VueJS devtools Chrome extension?

**Reply by emahuni** [reply]
See above for detailed method of getting devtools, but for hot reload when developing extensions there is an extension that purpotedly helps with this in the marketplace, it's called directus-extension-dev-canvas. I just noted it but never tried it though.
